Mathlib/RingTheory/PicardGroup.lean

```diff
@@ -0,0 +1,425 @@
+
+section Semiring
+
+variable [Semiring A] [Algebra R A] [FaithfulSMul R A]
+
+open LinearMap in
+instance projective_unit (I : (Submodule R A)ˣ) : Projective R I := by
+  obtain ⟨T, T', hT, hT', one_mem⟩ := mem_span_mul_finite_of_mem_mul (I.inv_mul ▸ one_le.mp le_rfl)
+  classical
+  rw [← Set.image2_mul, ← Finset.coe_image₂, mem_span_finset] at one_mem
+  set S := T.image₂ (· * ·) T'
+  obtain ⟨r, hr⟩ := one_mem
+  choose a ha b hb eq using fun i : S ↦ Finset.mem_image₂.mp i.2
+  let f : I →ₗ[R] S → R := .pi fun i ↦ (LinearEquiv.ofInjective
+      (Algebra.linearMap R A) (FaithfulSMul.algebraMap_injective R A)).symm.comp <|
+    restrict (mulRight R (r i • a i)) fun x hx ↦ by
+      rw [← one_eq_range, ← I.mul_inv]; exact mul_mem_mul hx (I⁻¹.1.smul_mem _ <| hT <| ha i)
+  let g : (S → R) →ₗ[R] I := .lsum _ _ ℕ fun i ↦ .toSpanSingleton _ _ ⟨b i, hT' <| hb i⟩
+  refine .of_split f g (LinearMap.ext fun x ↦ Subtype.ext ?_)
+  simp only [f, g, lsum_apply, comp_apply, sum_apply, toSpanSingleton_apply, proj_apply, pi_apply]
+  simp_rw [restrict_apply, mulRight_apply, id_apply, coe_sum, coe_smul, Algebra.smul_def,
+    ← Algebra.coe_linearMap, LinearEquiv.coe_toLinearMap, LinearEquiv.ofInjective_symm_apply,
+    mul_assoc, Algebra.coe_linearMap, ← Algebra.smul_def, ← Finset.mul_sum, eq,
+    (Finset.sum_coe_sort ..).trans hr.2, mul_one]
+
+theorem projective_of_isUnit {I : Submodule R A} (hI : IsUnit I) : Projective R I :=
+  projective_unit hI.unit
+
+end Semiring
+
+section CommSemiring
+
+variable [CommSemiring A] [Algebra R A] [FaithfulSMul R A]
+  (S : Submonoid R) [IsLocalization S A] (I J : (Submodule R A)ˣ)
+
+/-- Given two invertible `R`-submodules in an `R`-algebra `A`, the `R`-linear map from
+`I ⊗[R] J` to `I * J` induced by multiplication is an isomorphism. -/
+noncomputable def tensorEquivMul : I ⊗[R] J ≃ₗ[R] I * J :=
+  .ofBijective (mulMap' ..) ⟨by
+    have := IsLocalization.flat A S
+    simpa [mulMap', mulMap_eq_mul'_comp_mapIncl] using
+      (IsLocalization.bijective_linearMap_mul' S A).1.comp
+        (Flat.tensorProduct_mapIncl_injective_of_left ..),
+    mulMap'_surjective _ _⟩
+
+/-- Given an invertible `R`-submodule `I` in an `R`-algebra `A`, the `R`-linear map
+from `I ⊗[R] I⁻¹` to `R` induced by multiplication is an isomorphism. -/
+noncomputable def tensorInvEquiv : I ⊗[R] ↑I⁻¹ ≃ₗ[R] R :=
+  tensorEquivMul S I _ ≪≫ₗ .ofEq _ _ (I.mul_inv.trans one_eq_range) ≪≫ₗ
+    .symm (.ofInjective _ (FaithfulSMul.algebraMap_injective R A))
+
+include S in
+theorem units_invertible : Module.Invertible R I := .left (tensorInvEquiv S I)
+
+end CommSemiring
+
+section CommRing
+
+open CommRing (Pic)
+
+variable (R A : Type*) [CommRing R] [CommRing A] [Algebra R A] [FaithfulSMul R A]
+  (S : Submonoid R) [IsLocalization S A]
+
+/-- The group homomorphism from the invertible submodules
+in a localization of `R` to the Picard group of `R`. -/
+noncomputable def unitsToPic : (Submodule R A)ˣ →* Pic R :=
```

Comment:
Could you add `@[simps!]` and check that the generated simp lemmas are correct?

Reply:
```
adding projection Submodule.unitsToPic_apply:
      ∀ (R : Type u_3) (A : Type u_4) [inst : CommRing R] [inst_1 : CommRing A] [inst_2 : Algebra R A]
        [inst_3 : FaithfulSMul R A] (S : Submonoid R) [inst_4 : IsLocalization S A] (I : (Submodule R A)ˣ),
        (unitsToPic R A S) I = toPic R ↥↑I 
```